### PR TITLE
[lld-macho] Improve robustness of ObjC category merging

### PR DESCRIPTION
**Pull request by alx32**
This patch enhances the robustness of lld's Objective-C category merging. Currently, the category merger assumes it can fully parse and understand the format of all categories in the input, triggering an assert if any invalid category data is encountered. 

This will end up causing asserts in certain rare corner cases that are difficult to reproduce in small test cases. The proposed changes modify the behavior so that if invalid category data is detected, category merging is skipped for that specific class and all other categories sharing the same base class. This approach allows the linker to continue processing other categories without failing entirely due to a single problematic input.

We also add a LIT test to where we corrupt category data and check that category merging for that class was skipped but the link was successful.

### CODE REVIEW

**Review by kyulee-com — APPROVED** [review]
LGTM. Given the possibility of invalid input to this pass, falling back to not merging seems like the right thing to do instead of asserting and crashing. By the way, does LD64 behave in the same way?

**Comment by alx32** [comment]
> LGTM. Given the possibility of invalid input to this pass, falling back to not merging seems like the right thing to do instead of asserting and crashing. By the way, does LD64 behave in the same way?

Not sure about LD64 behavior. The issue here is very specific to the internals of LLD pointing to a StringPiece rather than a Symbol. LD64's implementation revolves around Atom's so the issue does not translate over.

**Comment by alx32** [comment]
Looking into `sed` failure for a quick forward fix.

**Comment by nico** [comment]
Looks like this breaks tests on Mac: http://45.33.8.238/macm1/94123/step_9.txt

(This is with the sed fix attempt.)

Please take a look and revert for now if it takes a while to fix.

**Comment by alx32** [comment]
> Please take a look and revert for now if it takes a while to fix.
Will revert in 1 hour if no fix by then.

**Comment by alx32** [comment]
> Looks like this breaks tests on Mac: http://45.33.8.238/macm1/94123/step_9.txt

Repro'ed on Mac + confirmed Fixed on Linux + Mac: https://github.com/llvm/llvm-project/pull/112981
I want to wait for validation to complete before merging because I switched the tool from sed to awk - or should I just merge right away ?

**Pull request merged**
